teps/0060-remote-resource-resolution.md

```diff
@@ -319,19 +324,252 @@ Why should this TEP _not_ be implemented?
 
 ## Alternatives
 
-<!--
-What other approaches did you consider and why did you rule them out?  These do
-not need to be as detailed as the proposal, but should include enough
-information to express the idea and why it was not acceptable.
--->
-
-## Infrastructure Needed (optional)
-
-<!--
-Use this section if you need things from the project/SIG.  Examples include a
-new subproject, repos requested, github details.  Listing these here allows a
-SIG to get the process for these resources started right away.
--->
+### Do Nothing
+
+Users wanting to refer to tasks in git (or anywhere else) are required to figure out their own
+implementation to fetch those resources, apply them, and use them in PipelineRuns / TaskRuns.
+
+Pros:
+- Pipelines has no work to do :tada:
+- Pipelines _could_ offer some supporting docs to help people do this.
+
+Cons:
+- No opportunity to standardize around remote fetching outside the sphere of tekton bundles.
+- If Pipelines did offer docs for people constructing their own solution then they'd need to be kept up to date.
+
+### Just Implement Git Support Directly in Pipelines
+
+Bake support for referencing tasks and pipelines in Git as a first class citizen alongside
+Tekton Bundles. Add a `git.Resolver` to live in tandem with [`oci.Resolver`](https://github.com/tektoncd/pipeline/blob/main/pkg/remote/oci/resolver.go).
+
+Pros:
+- Dramatically simpler requirements than an entire remote resource resolution proposal.
+- Way less specification noise, maintenance burden, etc.
+- The only other source of tekton resources that we've had actual requests for.
+
+Cons:
+- TODO. I'm not sure what these are.
+
+### Controller makes HTTP request to resolver to get resources
+
+Resolvers supporting different source types (git/oci/etc..) are registered with
+the Pipelines controller. The controller makes web requests to resolve remote resources:
+
+1. User submits a run type with a remote reference:
+
+```yaml
+kind: TaskRun
+metadata:
+  name: foo-run
+spec:
+  taskRef:
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    remoteType: git
+    remoteParams:
+    - name: url
+      value: https://github.com/tektoncd/catalog.git
+    - name: ref
+      value: main
+    - name: path
+      value: /tasks/git-clone/0.4/git-clone.yaml
+```
+
+2. Pipelines sends a webhook request to an endpoint registered for `git` remoteType.
+
+```json
+{
+  "url": "https://github.com/tektoncd/catalog.git",
+  "ref": "main",
+  "path": "/tasks/git-clone/0.4/git-clone.yaml"
+}
+```
+
+3. Endpoint resolves `git-clone.yaml` and returns its content in the response.
+
+```json
+{
+  "resolved": "apiVersion: tekton.dev/v1beta1\nkind: Task\nmetadata:\n  name: git-clone\nspec:\n [...]"
+}
+```
+
+Pros:
+- Writing an HTTP server that responds to requests with JSON is a relatively small development burden.
+- Possibly familiar pattern for anyone who's used Triggers' Interceptors?
```

Comment:
Interceptors have a CRD attached to them, and I think that's what we would have to do if went for this option.
We define a resolver CRD, that includes links to any secret that might be required.

I'd say that users cannot embed a resolver in the `TaskRun` or `PipelineRun`, they must rely on existing ones, which gives more control / security to operators. Resolvers however accept parameters (like for instance `branch` for a git provider CRD).

Reply:
Interesting, if multiple teams sharing a tekton cluster have their own repo of tasks would they each get their own instance of a resolver CRD?

Reply:
Uhm, good question - the moment we go for an HTTP based service, we break away from k8s RBAC, which does not make things easy for multi-tenancy. Indeed the triggers interceptors are defined cluster-wide. Task resolvers defined this way would not easily support multi-tenancy, unless we embed some auth layer in the HTTP service, but I'd rather not go there :) Maybe worth adding as a cons of this approach?

Reply:
Good idea, I've captured these as questions in the doc.